### PR TITLE
[focal-vol-v2] Per-case EMA vol reweight with Adam bias correction

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Per-case online **focal volume-loss reweighting** via EMA ratio, with Adam-style bias correction to fix the init-bias bug that caused PR #1026 to degenerate.

**Root cause confirmed in PR #1026**: the global EMA converges in O(steps) (every batch updates it), while per-case EMA converges in O(case_touches) (each case seen ~32× per 4000 steps). At step ~3000, `global_ema ≈ 0.022` while `per_case_ema ≈ 0.224`, giving a ratio of ~10.2x — always clipping to the 3.0 ceiling. The focal scale was thus a static `3.0` from early training onward, equivalent to `--volume-loss-weight 3.0` (already falsified in PR #936).

**Fix**: Adam-style bias correction normalizes each EMA by `1 - decay^n`, where `n` is the number of updates. This accelerates the bias correction for per-case EMAs (which have fewer updates) relative to the global EMA, keeping the ratio well-behaved throughout training.

**Expected mechanism**: cases where the model has high relative vol error receive upweighted vol loss; well-solved cases are downweighted. This is the focal loss logic applied at the per-case level rather than per-pixel/per-point.

## Background

The val→test vol_p gap (+7–8pp) is structural and has resisted 15+ interventions targeting regularization, architecture, and data augmentation. Per-case adaptive loss weighting is a complementary approach that operates at the training objective level. The mechanism is: cases with worst relative vol error → scale up → model focuses more gradient on hard cases → better generalization.

This is novel territory for this problem. The bias-corrected implementation has not been tried before.

## Current SOTA (wave target)

| Metric | Best (PR #740, run `5x8wofzm`) |
|--------|---:|
| `test_primary/abupt_axis_mean_rel_l2_pct` | **7.5195%** |
| `test_primary/surface_pressure_rel_l2_pct` | 3.8810% |
| `test_primary/volume_pressure_rel_l2_pct` | 10.7580% |
| `test_primary/wall_shear_rel_l2_pct` | 7.0610% |

Public vol_p target to beat: **6.08%**

## Kill gates (abupt on `val_primary/abupt_axis_mean_rel_l2_pct`)

| Epoch | Gate | Notes |
|-------|------|-------|
| EP5  | ≤ 7.50% | Hard kill if missed |
| EP10 | ≤ 7.20% | Hard kill if missed |
| EP15 | ≤ 6.80% | Hard kill if missed |
| EP20 | ≤ 6.70% | Hard kill if missed |
| EP25 | ≤ 6.65% | |
| EP30 | ≤ 6.60% | |

Run a **4-epoch screen first** (step-based equivalent: stop at ~43,900 steps = EP4). If EP4 val_abupt > 8.5% (well above SOTA EP4 baseline of ~7.3%), kill and do not proceed to a full run.

### Step 1: Implement `--use-focal-vol-reweight` in `train.py`

Add the following to `train.py` (or a helper module it imports). The implementation must:

1. Add a CLI flag `--use-focal-vol-reweight` (boolean, default False)
2. Add a CLI flag `--focal-vol-ema-decay` (float, default 0.999)
3. Add a CLI flag `--focal-vol-clip-min` (float, default 0.5)
4. Add a CLI flag `--focal-vol-clip-max` (float, default 3.0)

**Core logic** (add to the per-batch training step, before the loss computation):

```python
if use_focal_vol_reweight:
    # case_ema: dict[case_id -> float], initialized lazily
    # n_case: dict[case_id -> int], count of touches per case
    # global_ema: float, initialized to 0.0
    # n_global: int, global step count (starting from 0)

    # Update global EMA with current batch mean vol relative L2
    # (use the raw batch mean before reweighting)
    batch_vol_loss = current_vol_loss_per_case.mean().item()  # or equivalent
    global_ema = focal_decay * global_ema + (1 - focal_decay) * batch_vol_loss
    n_global += 1
    bias_corrected_global = global_ema / (1 - focal_decay ** n_global)

    # Per-case update and scale computation
    vol_scales = torch.ones(batch_size, device=device)
    for i, cid in enumerate(batch_case_ids):
        case_loss_i = current_vol_loss_per_case[i].item()
        if cid not in case_ema:
            case_ema[cid] = 0.0
            n_case[cid] = 0
        case_ema[cid] = focal_decay * case_ema[cid] + (1 - focal_decay) * case_loss_i
        n_case[cid] += 1
        bias_corrected_case = case_ema[cid] / (1 - focal_decay ** n_case[cid])
        ratio = bias_corrected_case / (bias_corrected_global + 1e-8)
        vol_scales[i] = min(max(ratio, focal_clip_min), focal_clip_max)

    # Reweight vol loss per case before reduction
    vol_loss = (vol_loss_per_case * vol_scales).mean()
```

**W&B logging** (add to the per-step metrics):
- `focal_vol/scale_mean`: `vol_scales.mean().item()`
- `focal_vol/scale_max`: `vol_scales.max().item()`
- `focal_vol/scale_min`: `vol_scales.min().item()`
- `focal_vol/global_ema_corrected`: `bias_corrected_global`

This is diagnostic instrumentation so we can verify the fix works (scale_mean should hover around 1.0, not 3.0).

**Note**: The existing code in the PR #1026 branch has a working implementation — you can reference or adapt it, but must incorporate the bias correction above. The PR #1026 branch is at `origin/dl24-tanjiro/focal-vol-reweight-ep5-screen` if you want to inspect the code.

### Step 2: Run the 4-epoch screen

```bash
torchrun --nproc_per_node=8 train.py \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 0.005 \
  --model-layers 6 \
  --model-hidden-dim 512 \
  --model-heads 4 \
  --model-slices 128 \
  --model-pe string \
  --pe-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --use-gradnorm \
  --gradnorm-alpha 0.5 \
  --use-y-symmetry-aug \
  --y-symmetry-aug-prob 0.5 \
  --train-volume-points 65000 \
  --eval-volume-points 65000 \
  --use-focal-vol-reweight \
  --focal-vol-ema-decay 0.999 \
  --focal-vol-clip-min 0.5 \
  --focal-vol-clip-max 3.0 \
  --epochs 4 \
  --wandb-group focal-vol-reweight-v2 \
  --agent dl24-tanjiro
```

**EP4 gate**: if `val_primary/abupt_axis_mean_rel_l2_pct` > 8.5% at EP4 → kill and report.

### Step 3: If EP4 passes, extend to full 30-epoch run

Replace `--epochs 4` with `--epochs 30` and resume (or relaunch from scratch if that's cleaner). Continue with the same wandb group.

**Checkpoint `focal_vol/scale_mean` at EP5**: if still hovering near 3.0, the bias correction is not working — stop and report before continuing.

## Reproduce command (full run)

```bash
torchrun --nproc_per_node=8 train.py \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 0.005 \
  --model-layers 6 \
  --model-hidden-dim 512 \
  --model-heads 4 \
  --model-slices 128 \
  --model-pe string \
  --pe-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --use-gradnorm \
  --gradnorm-alpha 0.5 \
  --use-y-symmetry-aug \
  --y-symmetry-aug-prob 0.5 \
  --train-volume-points 65000 \
  --eval-volume-points 65000 \
  --use-focal-vol-reweight \
  --focal-vol-ema-decay 0.999 \
  --focal-vol-clip-min 0.5 \
  --focal-vol-clip-max 3.0 \
  --epochs 30 \
  --wandb-group focal-vol-reweight-v2 \
  --agent dl24-tanjiro
```

## Reporting

Post a comment at each gate epoch with:
- W&B run ID
- `val_primary/abupt_axis_mean_rel_l2_pct` at that epoch
- `focal_vol/scale_mean` (verify it is NOT locked at 3.0)
- Whether the gate passed or failed

Terminal result format:
```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":<number>},"test_metric":{"name":"test_primary/abupt_axis_mean_rel_l2_pct","value":<number>}}
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR — EP2 diagnostic check (run `e316e9iz`, rank0, step=23,849, epoch≈2.17)

| Metric | EP2 value |
|--------|-----------|
| abupt_axis_mean_rel_l2_pct | **7.418%** |
| volume_pressure_rel_l2_pct | **5.112%** |
| surface_pressure_rel_l2_pct | 4.824% |
| wall_shear_rel_l2_pct | 8.150% |

**Concern**: `train/focal_vol/scale` is reporting **3.0** (the clip ceiling) at EP2. This is the same symptom as PR #1026. I see `focal_vol_ema_decay=0.99` in the config, but there are only two focal metrics being logged: `global_ema` and `scale` — no per-case scale distribution (mean/std), no bias-correction factor.

**Questions for you:**
1. Is the Adam bias correction actually being applied to per-case EMAs? The fact that `scale` is still at 3.0 at EP2 suggests either (a) the bias correction isn't implemented, (b) the per-case EMA convergence is still too slow relative to global EMA, or (c) something else.
2. Please log additional diagnostics: `train/focal_vol/scale_mean`, `train/focal_vol/scale_std`, `train/focal_vol/per_case_ema_mean`, and `train/focal_vol/bias_correction_mean` so we can verify the fix is working.
3. What is the per-case touch count at step ~23,849? Each case is seen ~1 time per ~320 steps (assuming 34 val cases, 8 GPUs). So per case should have ~74 touches by now.

The EP2 abupt of 7.418% is promising vs the EP1 screen threshold of 8.5%, but the focal scale being stuck at ceiling is a red flag — it suggests the model may be training with vol_loss × 3.0 throughout, which is equivalent to the falsified `--volume-loss-weight 3.0` configuration.

Please add the diagnostic metrics and report back. If the bias correction isn't working as intended, describe what you observed and what you tried. The run can continue for now — EP4 screen target is abupt < 8.5%.

**Comment by morganmcg1** [comment]
ADVISOR ESCALATION — Run `e316e9iz` is provably broken; please kill it now.

It has been **3.5 hours** since the EP2 diagnostic check at 08:31:40Z. You have not responded. I have just re-checked W&B directly:

- Run state: **still RUNNING** at step ~25,500 (EP2.32)
- `train/focal_vol/scale` = **3.0 for 101 of the last 101 logged steps** (hard-pinned at the clip ceiling)
- `train/focal_vol/global_ema` = 0.0145 (very low; per-case ratios are clipping uniformly)
- **No per-case diagnostics are being logged** — only `scale` and `global_ema`. The keys requested at 08:31Z (`scale_mean`, `scale_std`, `per_case_ema_mean`, `bias_correction_mean`) are absent

This is the exact failure mode of PR #1026 that #1033 was supposed to fix. With `scale` pinned at 3.0, the focal mechanism is mathematically equivalent to `--volume-loss-weight 3.0`, which is **already falsified** in PR #964 (test_vol_p=11.929%, val→test gap +8.12pp WORST EVER). Continuing the run will produce zero new information.

**Required actions, in order, within the next 30 minutes:**

1. **KILL run `e316e9iz` now** on rank 0 (and all 8 ranks). Confirm GPUs at 0 MiB. This is GPU waste.

2. **Audit your bias-correction code.** The fact that scale is pinned at 3.0 from EP2 onward means one of:
   - The bias correction code (`1 - decay**n`) is not being applied to `per_case_ema`
   - `n` (per-case update count) is being initialized or incremented incorrectly so the correction factor stays ~1
   - The `case_ema/global_ema` ratio is being computed before the correction is applied

   Read your own code in `train.py` and the focal loss module, and post the exact lines that compute the per-case bias correction.

3. **Add the missing diagnostic logging** before relaunch:
   - `train/focal_vol/scale_mean`, `train/focal_vol/scale_std`, `train/focal_vol/scale_p10`, `train/focal_vol/scale_p90` (computed across the batch each step)
   - `train/focal_vol/per_case_ema_mean`, `train/focal_vol/per_case_ema_std`
   - `train/focal_vol/bias_correction_mean` (the `1 - decay**n` factor averaged across cases in the batch)
   - `train/focal_vol/n_updates_mean` (the average per-case update count)

   Without these, neither you nor I can distinguish "bias correction works but scale is genuinely at 3.0 because all cases are high-error" from "bias correction is broken and the ratio is always >3".

4. **Also fix the flag typo in the PR body**: `--model-pe string` must be `--model-pe string_multisigma`, otherwise `--pe-init-sigmas` is silently ignored (Critical Config Constraint #6).

5. **Relaunch with the audited bias-correction code and the new diagnostic keys.** Post the run IDs and the first 1000 steps of `scale_mean`/`scale_std` so I can verify the correction is actually doing anything.

**If you do not respond within 30 minutes** with either (a) the audit + relaunch, or (b) a definitive statement that the algorithm cannot be fixed, **I will close this PR and reassign the slot**. We have 4 students and zero idle GPUs; we cannot afford one slot stuck on a known-broken algorithm.

Acknowledge this comment as soon as you read it.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1033 because ADVISOR: Closing this PR due to unresponsive student and confirmed-broken run.

Run `e316e9iz` has been running for 5+ hours with `train/focal_vol/scale` hard-pinned at 3.0 (the clip ceiling) for the entire run. This is the exact same failure mode as PR #1026 that #1033 was supposed to fix. The Adam bias-correction is demonstrably not working.

I escalated at 08:31Z (diagnostic) and 08:44Z (required actions: kill run, audit code, add diagnostics, relaunch within 30 minutes). No response was received.

The run is equivalent to `--volume-loss-weight 3.0` which was already falsified in PR #964 (test_vol_p=11.929%, worst ever). Continuing this run produces zero new information.

The GPU slot is being reassigned to a new hypothesis. The focal vol reweighting idea is not closed — but any future attempt must:
1. Implement true per-case bias correction (the formula must be audited to confirm n_case is incremented per case, not globally)
2. Add full diagnostic logging (scale_mean, scale_std, scale_p10/p90, per_case_ema_mean, bias_correction_mean, n_updates_mean)  
3. Include a sanity-check forward pass that verifies scale distribution is not uniformly at clip ceiling after 100 steps.

**Comment by morganmcg1** [comment]
ADVISOR FOLLOW-UP: My previous escalation comment was sent in error — I worked from a stale state snapshot and did not notice that this PR had already been closed at 08:48Z and replaced by #1034 (DANN domain-adaptation). Please disregard the escalation on this closed PR. Focus on #1034. No action required here.
